objc/objc_runtime_darwin.go

```diff
@@ -148,203 +168,169 @@ func AllocateClassPair(super Class, name string, extraBytes uintptr) Class {
+// The name also cannot contain any spaces.
+// The type is the Go equivalent type of the Ivar.
+// Attribute determines if a getter and or setter method is generated for this field.
+type FieldDef struct {
+	Name      string
+	Type      reflect.Type
+	Attribute IvarAttrib
+}
+
+// The RegisterClass function takes the name of the class to create, the superclass, a list of protocols this class
+// implements, a list of fields this class has and a list of methods. It returns the created class or an error
+// describing what went wrong.
+func RegisterClass(name string, superClass Class, protocols []*Protocol, ivars []FieldDef, methods []MethodDef) (Class, error) {
+	class := objc_allocateClassPair(superClass, name, 0)
 	if class == 0 {
-		return 0, fmt.Errorf("objc: failed to create class with name '%s'", split[0])
+		return 0, fmt.Errorf("objc: failed to create class with name '%s'", name)
 	}
-	if len(split) > 2 {
-		// Add Protocols
-		for _, n := range split[2:] {
-			if !class.AddProtocol(GetProtocol(n)) {
-				return 0, fmt.Errorf("objc: couldn't add Protocol %s", n)
-			}
+	// Add Protocols
+	for _, p := range protocols {
+		if !class.AddProtocol(p) {
+			return 0, fmt.Errorf("objc: couldn't add Protocol %s", protocol_getName(p))
 		}
 	}
-	// Add exported methods based on the selectors returned from Selector(string) SEL
-	for i := 0; i < ptr.NumMethod(); i++ {
-		met := ptr.Method(i)
-		// we know this method is the interface one since RegisterClass
-		// requires that the struct implement Selector.
-		if met.Name == "Selector" {
-			continue
-		}
-		sel := object.Selector(met.Name)
-		if sel == 0 {
-			continue
-		}
-		fn := met.Func.Interface()
+	// Add exported methods based on the selectors returned from ClassDef(string) SEL
+	for idx, def := range methods {
 		imp, err := func() (imp IMP, err error) {
 			defer func() {
 				if r := recover(); r != nil {
 					err = fmt.Errorf("objc: failed to create IMP: %s", r)
 				}
 			}()
-			return NewIMP(fn), nil
+			return NewIMP(def.Fn), nil
 		}()
 		if err != nil {
-			return 0, fmt.Errorf("objc: couldn't add Method %s: %w", met.Name, err)
+			return 0, fmt.Errorf("objc: couldn't add Method at index %d: %w", idx, err)
 		}
-		encoding, err := encodeFunc(fn)
+		encoding, err := encodeFunc(def.Fn)
 		if err != nil {
-			return 0, fmt.Errorf("objc: couldn't add Method %s: %w", met.Name, err)
+			return 0, fmt.Errorf("objc: couldn't add Method at index %d: %w", idx, err)
 		}
-		if !class.AddMethod(sel, imp, encoding) {
-			return 0, fmt.Errorf("objc: couldn't add Method %s", met.Name)
+		if !class.AddMethod(def.Cmd, imp, encoding) {
+			return 0, fmt.Errorf("objc: couldn't add Method at index %d", idx)
 		}
 	}
 	// Add Ivars
 	// Start at 1 because we skip the class object which is first
-	for i := 1; i < strct.NumField(); i++ {
-		f := strct.Field(i)
-		if f.Name == "_" {
-			continue
+	for _, ivar := range ivars {
+		if strings.Contains(ivar.Name, " ") {
```

Comment:
Let's check the name with a regexp to avoid future troubles like this regexp expression:

```
[a-z][a-zA-Z0-9]*
```

As far as I know, the first character is always a small cap.

FYI, even if an ivar name starts with an abbreviation, the setter name is always like setFooBar, not setFOOBar. For example, Metal has `setRgbBlendOperation` for `rgbBlendOperation`, instead of `setRGBBlendOperation`. So the current implemention for setters seeems fine to me.